File: src/python/battle_state.py
class BattleStateTracker:
    def __init__(self):
        self.battle_history = {}
    
    def update(self, battle):
        battle_id = battle.battle_tag
        if battle_id not in self.battle_history:
            self.battle_history[battle_id] = []
    
        current_state = {
                "turn": battle.turn,
                "active_pokemon": {
                    "self": self._extract_pokemon_data(battle.active_pokemon),
                    "opponent": self._extract_pokemon_data(battle.opponent_active_pokemon)
                },
                "team": {
                    "self": {p.species: self._extract_pokemon_data(p) for p in battle.team.values()},
                    "opponent": {p.species: self._extract_pokemon_data(p) for p in battle.opponent_team.values() if p.species}
                },
                "weather": battle.weather,
                "fields": battle.fields
            }
        
        self.battle_history[battle_id].append(current_state)
        return current_state
    
    def _extract_pokemon_data(self, pokemon):
        if not pokemon:
            return None
        
        return {
            "species": pokemon.species,
            "hp": pokemon.current_hp_fraction,
            "max_hp": pokemon.max_hp,
            "level": pokemon.level,
            "status": pokemon.status,
            "types": pokemon.types,
            "moves": [move.id for move in pokemon.moves.values()] if hasattr(pokemon, 'moves') else [],
            "stats": pokemon.stats if hasattr(pokemon, 'stats') else {}
        }

File: src/csharp/PokemonDashboard/Models/BattleState.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace PokemonDashboard.Models
{
    public class BattleState
    {
        public ActivePokemonPair? ActivePokemon { get; set; }
        public Dictionary<string, Dictionary<string, double>>? PayoffMatrix { get; set; }
        public Dictionary<string, double>? MoveProbabilities { get; set; }
        public int Turn { get; set; }
        public string? Weather { get; set; }
        public List<string>? Fields { get; set; }
    }
    
    public class ActivePokemonPair
    {
        public Pokemon? Self { get; set; }
        public Pokemon? Opponent { get; set; }
    }
    
    public class Pokemon
    {
        public string? Species { get; set; }
        public double Hp { get; set; }
        public List<string>? Types { get; set; }
        public List<Move>? Moves { get; set; }
    }
    
    public class Move
    {
        public string? Id { get; set; }
        public string? Name { get; set; }
        public string? Type { get; set; }
        public int BasePower { get; set; }
        public int Category { get; set; }
    }
}